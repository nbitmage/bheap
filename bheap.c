#include "bheap.h"
#include <stdio.h>
#include <stdlib.h>

#define TRUE 1
#define FALSE 0
#define SWAP(type,a,b)  {type work = a; a = b; b = work;}

void bheap_remove_root(bheap*);
void bheap_bubble_up(bheap*);
void bheap_bubble_down(bheap*);
int bheap_extend(bheap*);

bheap*
bheap_new(int len)
{
  bheap *heap;

  heap = (bheap*)malloc(sizeof(bheap));
  heap->nodes = (bheap_node*)malloc(sizeof(bheap_node) * len);
  heap->len = len;
  heap->tail = -1;

  return heap;
}

void
bheap_free(bheap* heap)
{
  free(heap->nodes);
  free(heap);
  return;
}

int
bheap_push(bheap* heap, int val, void* opt)
{
  bheap_node *node;
  // if (heap->tail == heap->len - 1) bheap_extend(heap);
  node = &heap->nodes[++heap->tail];
  node->val = val;
  node->opt = opt;
  bheap_bubble_up(heap);
  return TRUE;
}

int
bheap_is_empty(bheap* heap)
{
  if (heap->tail < 0) {
    return TRUE;
  } else {
    return FALSE;
  }
}

void
bheap_remove_root(bheap* heap)
{
  heap->nodes[0] = heap->nodes[heap->tail--];
  bheap_bubble_down(heap);
}

bheap_node
bheap_pop(bheap* heap)
{
  bheap_node node;
  node = heap->nodes[0];
  bheap_remove_root(heap);
  return node;
}

void
bheap_print(bheap* heap)
{
  int i;
  for (i = 0; i <= heap->tail; i++) {
    printf("%d\n", heap->nodes[i].val);
  }
}

void
bheap_bubble_up(bheap* heap)
{
  int i;
  bheap_node *nodes, tail_node;
  nodes = heap->nodes;
  tail_node = nodes[heap->tail];

  for (i = heap->tail; i > 0 && tail_node.val > nodes[i/2].val; i = i / 2) {
    nodes[i] = nodes[i/2];
  }
  nodes[i] = tail_node;
}

void
bheap_bubble_down(bheap* heap)
{
  int parent, child;
  bheap_node *nodes, root_node;
  nodes = heap->nodes;
  root_node = nodes[0];

  parent = 0;
  while (parent <= heap->tail) {
    child = (parent * 2) + 1;
    if (child > heap->tail) break;
    if (child + 1 <= heap->tail && nodes[child].val < nodes[child+1].val)
      child++;
    if (root_node.val < nodes[child].val) {
      nodes[parent] = nodes[child];
      parent = child;
    } else break;
  }
  nodes[parent] = root_node;
}
