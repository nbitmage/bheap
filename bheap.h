#ifndef _BINARY_HEAP_LIB_H_
#define _BINARY_HEAP_LIB_H_

typedef struct {
  int val;
  void* opt;
} bheap_node;

typedef struct {
  bheap_node *nodes;
  int len;
  int tail;
} bheap;

bheap* bheap_new(int);
void bheap_free(bheap*);

int bheap_add(bheap*, int, void*);
bheap_node bheap_get(bheap*);
int bheap_is_empty(bheap*);

void bheap_print(bheap* heap);

#endif
