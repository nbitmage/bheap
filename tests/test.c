#include "bheap.h"
#include <stdio.h>

typedef struct {
  int id;
  int price;
  char* name;
} item_t;

int main(void){
  bheap *h;
  item_t *p;

  // prepare data for test
  item_t item1 = {8,  120,  "apple"};
  item_t item2 = {4,  80,   "orange"};
  item_t item3 = {32, 100,  "banana"};
  item_t item4 = {2,  130,  "kiwi"};
  item_t item5 = {8,  150,  "grape"};

  // create binary-heap structure
  h = bheap_new(10);
  // add data
  bheap_push(h, item1.id, (void*)&item1);
  bheap_push(h, item2.id, (void*)&item2);
  bheap_push(h, item3.id, (void*)&item3);
  bheap_push(h, item4.id, (void*)&item4);
  bheap_push(h, item5.id, (void*)&item5);

  // iterate process until the heap become empty
  while (!bheap_is_empty(h)) {
    // get data in the root of the heap.
    p = (item_t*)bheap_pop(h).opt;
    printf("id:%d,name:%s\n", p->id, p->name);
  }

  // free memory space allocated for heap
  bheap_free(h);
  return 0;
}
