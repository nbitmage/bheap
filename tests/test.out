id:32,name:banana
id:8,name:apple
id:8,name:grape
id:4,name:orange
id:2,name:kiwi
